atl-checker/src/algorithms/certain_zero/search_strategy/linear_programming_search.rs

```diff
@@ -0,0 +1,224 @@
+                self.queue.push(edge, -dist);
+            } else {
+                // Todo what should default value be, if cannot be calculated? For now: first priority
+                self.queue.push(edge, 0);
+            }
+        }
+    }
+}
+
+impl LinearProgrammingSearch {
+    /// Finds the distance of an edge using linear programming
+    fn get_distance_of_edge(&mut self, edge: &Edge<AtlVertex>) -> Option<i32> {
+        let source = edge.source();
+
+        // If we have seen this source before, get the result instantly
+        if let Some(sol) = self.result_cache.get(&source) {
+            return Some(sol.1);
+        }
+
+        // If we have not seen this formula before, calculate constrained phi
+        if !self.phi_cache.contains_key(&source.formula()) {
+            // Convert the formula to a structure of constraints
+            let lcp = self.phi_mapper.map(&self.game, &*source.formula());
+            self.phi_cache.insert(source.formula(), lcp);
+        }
+
+        // Get constraints of this phi
+        let constrained_phi = self.phi_cache.get(&source.formula()).unwrap();
+        // Get current state in vertex
+        let state = self.game.state_from_index(source.state());
+
+        // Use linear programming to find closest state that satisfies the constraints
+        if let Some(sol) = self.get_optimal_distance(constrained_phi, &state) {
+            let dist = sol.1;
+            // Cache the resulting solution
+            self.result_cache.insert(source.clone(), sol);
+            return Some(dist);
+        }
+        // If we could not find a distance, return None
+        None
+    }
+
+    fn get_optimal_distance(
+        &self,
+        constrained_phi: &LinearConstrainedPhi,
+        state: &State,
+    ) -> Option<(State, i32)> {
+        // The LinearConstrainedPhi contains multiple variants of the linear problem due to the ORs.
+        // So we iterate through them and find the best result
+        let mut best: Option<(State, i32)> = None;
+        for constraints in all_variants(constrained_phi) {
+            // We keep track of the symbols/variables encountered
+            let mut symbol_vars: HashMap<SymbolIdentifier, Variable> = HashMap::new();
+            let mut problem = Problem::new(OptimizationDirection::Minimize);
+
+            for constraint in &constraints {
+                if constraint.comparison == ComparisonOp::NotEqual {
+                    // We skip not-equal constraints since the minilp cannot handle them.
+                    // They don't restrict the solution space in a meaning way anyway.
+                    continue;
+                }
+                // Build constraint
+                let mut lin_expr = LinearExpr::empty();
+                // The contraint is offset by the given state. This is reflected by adding
+                // coefficient*value_of_symbol_in_state to the constant
+                let mut offset = -constraint.constant;
```

Comment:
I'm not quite sure whats going on here

Reply:
This is where we take the state into account. First of all, note that minilp expects the constant on the RHS instead of LHS like our LinearConstraint struct, which is why the constant is negative.

Now if the constraint is `ax+by+cz<k` and we want to find the closest solution to the state `(x0, y0, z0)`, then we have to translate (move) the constraint by subtracting the state, such our the state becomes the origin. This becomes `a(x-x0)+b(y-y0)+c(z-z0)<k` => `ax-ax0+by-by0+cz-cz0<k` => `ax+by+cz<k+ax0+by0+cz0`.

Reply:
![image](https://user-images.githubusercontent.com/33687173/118376079-16774780-b5c6-11eb-828c-388282b234b4.png)

So like this? (circle is our state)

Reply:
Yes. So ATL vertices with similar formula might have different solutions because the state is very different.